[SYCL][Graph] Add spec wording for graph-owned memory allocations (#384)
reble/llvm: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

update API, then any previous updates to the dynamic parameter will be reflected
in the new graph.

==== Graph-Owned Memory Allocations [[graph-memory-allocations]]

It can be desirable for a graph to own and manage memory allocations for memory
associated with commands in the graph. This is made possible by using the
`async_malloc` and `async_free` functions from the

> link:../experimental/sycl_ext_codeplay_async_memory_alloc.asciidoc[sycl_ext_codeplay_async_memory_alloc

Reviewer: oneapi extension not codeplay, and in a few other places

Author: Fixed thanks!

] extension. These commands can be added to a graph either via queue recording
or explicit graph creation, which will create allocations which are owned and
managed by that specific `command_graph`.

This can be useful for example when capturing library calls, as the library may
require some temporary memory for commands they record to the graph. Use of
these APIs by the library enables the creation and management of these
allocations within the graph, without the library needing to manage their

> lifetime or be aware that the commands are being recorded to a graph.

Reviewer: I'm not sure about this, but is this paragraph something that could/should be in the "recording library calls" section of the usage guide?

Author: I was also unsure about including this here, I think you're right it could be moved. I think this advice is more for library authors than applications recording library calls though, so I might make a new section for that.

Author: Have made this change and expanded it a bit to talk about the allocator use case and providing this as an alternative for graph compatibility.

same way as regular USM pointers.

===== Supported Features

Currently only device allocations are supported in graphs. Attempting to
add allocations of any other type to a graph will result in synchronous errors
being thrown.

===== Restrictions

The following restrictions apply to any graph using these allocations:

* The graph cannot be finalized more than once.
* The graph cannot be used as a sub-graph in another graph.
* Graph memory allocation nodes cannot be updated, and graphs containing these
nodes cannot be updated via <<whole-graph-update, Whole Graph Update>>.

===== Behaviour 

The semantics of `async_malloc` and `async_free` within a graph differ from the
eager SYCL usage described in the
link:../experimental/sycl_ext_codeplay_async_memory_alloc.asciidoc[sycl_ext_codeplay_async_memory_alloc]
extension.

* Graph memory allocations are not made directly from any default or
user-provided Memory Pool. Each graph containing alloc/free nodes maintains its
own pool of memory from which allocations are made. The following properties
of a default or user-provided memory pool provided in calls to `async_malloc`

> will be respected for the associated graph allocations:

Reviewer: should way say explicitly that any other properties will be ignored?

Author: Good point, I've done this and also added that it will respect the allocation kind of the pool as that was missing before.

update API, then any previous updates to the dynamic parameter will be reflected
in the new graph.

==== Graph-Owned Memory Allocations [[graph-memory-allocations]]

> It can be desirable for a graph to own and manage memory allocations for memory

Reviewer 2: I think we could also mention here that this allows the lifetime of memory allocations to be automatically tied to the lifetime of a graph.

Author: I've mentioned this in this initial intro section and also moved the "Allocation Lifetimes" section closer to the top which goes into a little more detail on valid usage.

associated with commands in the graph. This is made possible by using the

> `async_malloc` and `async_free` functions from the

Reviewer 2: ```suggestion
`async_malloc` and `async_free` commands from the
```

link:../experimental/sycl_ext_oneapi_async_memory_alloc.asciidoc[sycl_ext_oneapi_async_memory_alloc

> ] extension. These commands can be added to a graph either via queue recording

Reviewer 2: ```suggestion
] extension. These commands can be added to a graph either via queue recording the async malloc and free enqueue functions
```

Author: Is this referring to the free functions that don't return events specifically? I'm not really sure of the need for this clarification here, given we have previously just mentioned the functions by name and the API usage section immediately below this gives examples of both the handler versions and the enqueue functions for both APIs.

Reviewer 2: So my thinking here was to clarify that the nodes are added by recording the `async_malloc` / `async_free` functions, but maybe your right and we don't need to be so explicit here.

or explicit graph creation, which will create allocations which are owned and

> managed by that specific `command_graph`.

Reviewer 3: From a structure point of view, it could make sense to give an example of how the API looks like in here (maybe by moving the API usage section). When reading this initially, I got a bit confused and had to scroll down to the API Usage section.

Author: That's fair, I've moved the API usage section to immediately after this introductory section, does that read better?

Reviewer 3: Thanks for the changes, I think it reads better now :+1:

> Graphs use vitual memory to manage these allocations. To use this feature in a

Reviewer 3: ```suggestion
Graphs use virtual memory to manage these allocations. To use this feature in a
```

> Graphs use vitual memory to manage these allocations. To use this feature in a

Reviewer 2: I don't think we should require this to use virtual memory, this should be an implementation detail.

> Graphs use virtual memory to manage these allocations. To use this feature in a

Reviewer 2: I think we should reiterate that the async malloc functions must return a valid USM pointer, but we shouldn't require that this is implemented using virtual memory, this should be an implementation detail.

I think we could keep a note about this, but I'd suggest moving this to an implementation details section, so it's non-normative.

Author: > I think we should reiterate that the async malloc functions must return a valid USM pointer

I don't disagree that virtual memory could be an implementation detail, but I think we can't guarantee this exactly as you describe. I think we would still need some wording that effectively says, "the returned pointer is not valid/safe for use outside of the graph until the graph has been finalized". Since an implementation using virtual memory probably doesn't want to allocate physical mem and map during graph building, the pointer is not actually valid for full use outside of a graph until that graph has been finalized.

Author: Wording about virtual memory has now been removed.

> graph the device associated with the graph must support the

Reviewer 3: ```suggestion
graph, the device associated with the graph must support the
```

`ext_oneapi_virtual_mem` aspect. Attempting to use this feature with devices

> that do not support this aspect will result in synchronous errors being thrown.

Reviewer 2: We should state here what kind of error is thrown here.

Author: Updated this to specify `feature_not_supported`, though it may get removed depending on the other discussions around virtual memory in the spec.

Pointers returned from allocation nodes can be used in other graph nodes in the
same way as regular USM pointers.

> 

Reviewer 2: We should also define the lifetime of memory allocations from async malloc/free nodes in a graph, that they will exist as long as the graph is alive.

===== Supported Features [[allocation-supported-features]]

Currently only device allocations are supported in graphs. Attempting to

> add allocations of any other type to a graph will result in synchronous errors

Reviewer 2: We should specify the kind of exceptions that are thrown.

Author: Changed to specifiy that a `feature_not_supported` error code will be thrown.

being thrown.

===== Restrictions

> The following restrictions apply to any graph using these allocations:

Reviewer 2: ```suggestion
The following restrictions apply to any graph containing async malloc or free nodes:
```

* The graph cannot be finalized more than once.
* The graph cannot be used as a sub-graph in another graph.

> * Graph memory allocation nodes cannot be updated, and graphs containing these

Reviewer 3: A bit off topic:

I undestand why this limitation to update exists. But I wonder if it will limit the practical usecases of this functionality. Is it common for applications to be happy to just reuse the same graph without ever needing to update it?

Author: As a consequence of not being able to update allocation nodes it does limit the ability to use whole-graph update specifically, but you can still use explicit graph update on other nodes in a graph which don't have this limitation.

Whole-graph update is problematic I think because the use of the function creates this promise/contract that all of the nodes in the executable graph will be updated to match the state of the source modifiable graph. We could skip allocation nodes during the update but that creates a problem since that modifiable graph will have different virtual addresses that are used in the other nodes in the graph. 

It also breaks that promise/contract which is somewhat unintuitive, since the state of the executable graph will not actually match the source graph.

* Graph memory allocation nodes cannot be updated, and graphs containing these
nodes cannot be updated via <<whole-graph-update, Whole Graph Update>>.

===== Behaviour 

The semantics of `async_malloc` and `async_free` within a graph differ from the
eager SYCL usage described in the
link:../experimental/sycl_ext_oneapi_async_memory_alloc.asciidoc[sycl_ext_oneapi_async_memory_alloc]
extension.

* Graph memory allocations are not made directly from any default or
user-provided Memory Pool. Each graph containing alloc/free nodes maintains its
own pool of memory from which allocations are made. The following properties of
a default or user-provided memory pool provided in calls to
`async_<malloc/malloc_from_pool>` will be respected for the associated graph
allocations, all other properties will be ignored:

** The allocation type specified when creating the pool with
`usm::alloc::<host/device/shared>`, subject to the limitations in the
<<allocation-supported-features, supported features>> section.

** `property::memory_pool::read_only` - Guarantee from the user that memory is
only being read from, the implementation may be able to optimize in this case.

> ** `property::memory_pool::zero_init` - Physical memory will be zero-initialized

Reviewer 2: I think we should clarify here whether the zero initialisation happens:
1. the first time the memory is allocated during the first execution of the graph, or
2. the first time the memory is allocated during each execution of the graph
My assumption here is that we would want the (1) here as that's most consistent with the non-graph behaviour.

Author: I have this specified here as only during the initial allocation in finalize, but perhaps then something more generic like "it will only be initialized once when the memory is initially allocated, and not on subsequent graph executions" is clearer/less restrictive to an implementation in this case?

on first allocation during the call to `command_graph::finalize()`. It will not
be zero-initialized again during execution of the graph. If that is required by
the application it is the responsibility of the user to add the appropriate
commands to the graph to do this.

> * `node_type::async_malloc` nodes within a graph will reserve a virtual address

Reviewer 2: I think this part should be left as implementation details, so we say that async malloc will provide a pointer which can be used in graph nodes between then and the free node, but not specify how or where from the memory is allocated.

There are cases where we may not be able to use virtual memory and would need to emulate this behaviour, so I think implementations will need to have flexibility here.

> During graph construction no physical memory is allocated on the device.

Reviewer 2: I think this part should be implementation detail.

Author: I've removed this wording now.

(VA) range on the device and return a pointer to the start of that VA range.
This pointer can then be used in other graph nodes until an `async_free` node is
added for that pointer. Returned pointers are not guaranteed to be unique. An
implementation may return the same pointer as a previous `async_malloc` node if
that pointer was previously freed via `async_free` at that point in the graph.

> * `node_type::async_free` nodes within a graph indicate that a given pointer/VA

Reviewer 2: Similar to above, I think we should leave how memory is freed as an implementation detail.

range is no longer in use, and may be available for subsequent re-use. No
de-allocation of physical memory on a device is done during calls to
`async_free`.

* Other nodes which use a given graph allocation must be ordered via
dependencies such that they are ordered after the alloc node and before the free
node for a given allocation. It is the user's responsibility to ensure that
dependencies are correct. Using a pointer in a graph command ordered after it
has been freed via an `async_free` node results in undefined behavior.

> The total amount of physical memory being used by a graph can be queried using

Reviewer 2: I think we may want to have a separate query or have different return value, depending on whether the graph is modifiable or finalised. When it's modifiable, this is before any optimisations are necessarily applied, so it would likely have to be the highest potential memory footprint for the graph. But then once it's finalised and optimisations for memory re-use have been applied, it can tell you the actual memory requirement on the native graph.

  Ptr = async_malloc(CGH, usm::alloc::device, AllocSize);
}

// Explicit graph creation
Graph.add(CGF);

Graph.add([&](handler &CGH){
  async_free(CGH, Ptr);
});

// Queue recording
Graph.begin_recording(Queue);
Queue.submit(CGF);
Queue.submit([&](handler &CGH){
  async_free(CGH, Ptr);
});
----

The `async_*` commands which take a queue can also be used with queue
recording, particularly when recording an in-order queue to specify dependencies
as no SYCL event is returned.

[source,c++]
----
void* Ptr = nullptr;
size_t AllocSize = 1024;
queue Queue {syclContext, syclDevice, {property::queue::in_order{}}};

Graph.begin_recording(Queue);
Ptr = async_malloc(Queue, usm::alloc::device, AllocSize);
async_free(Queue, Ptr);
----

===== Supported Features [[allocation-supported-features]]

Currently only device allocations are supported in graphs. Attempting to
add allocations of any other type to a graph will result in synchronous errors
being thrown with error code `feature_not_supported`.

===== Restrictions

The following restrictions apply to any graph containing async malloc or free
nodes:

* Only one executable graph instance (created by finalizing the modifiable

> graph) can be alive at any time, and must be destroyed before the graph can be

Reviewer 2: I think we should mention the reference semantics here in the normative wording, saying something to the effect of all copies of the executable graph must be destroyed before the modifiable graph and be finalised again.

> finalized again.

Reviewer 2: We should state what happens in the case of attempting to finalise a modifiable graph containing malloc/free nodes when an executable graph is still alive.

Author: Added wording about throwing an error on finalize here, it was already defined in the finalize function definition itself.

own pool of memory from which allocations are made. The following properties of
a default or user-provided memory pool provided in calls to
`async_<malloc/malloc_from_pool>` will be respected for the associated graph
allocations, all other properties will be ignored:

** The allocation type specified when creating the pool with
`usm::alloc::<host/device/shared>`, subject to the limitations in the
<<allocation-supported-features, supported features>> section.

** `property::memory_pool::read_only` - Guarantee from the user that memory is
only being read from, the implementation may be able to optimize in this case.

** `property::memory_pool::zero_init` - Allocated memory will be
zero-initialized only once when first allocated. It will not be zero-initialized
again before or during any subsequent executions of the graph. If that is
required by the application it is the responsibility of the user to add the
appropriate commands to the graph to do this.

* `node_type::async_malloc` nodes within a graph will return a pointer to an
allocation of the provided size. This pointer can then be used in other graph
nodes ordered after that node in the same way any USM pointer would be. 

> NOTE: Returned pointers are not guaranteed to be unique. An

Reviewer 2: We've started using this format for non-normative notes in extensions:

```
[Note: ... -- end note]
```

  property for more information.
* Throws with error code `invalid` if the type of the command-group is not a
  kernel execution and a `dynamic_parameter` was registered inside `cgf`.
* Throws with error code `invalid` if the type of the command-group is
`async_malloc` and the `usm::alloc` type of the associated memory pool is not

> `usm::alloc::device

Reviewer: ```suggestion
`usm::alloc::device`.
```

|Synchronous operation that creates a new graph in the executable state with a
fixed topology that can be submitted for execution on any queue sharing the

> context associated with the graph. It is valid to call this method multiple

Reviewer 2: ```suggestion
context associated with the graph. It is valid to call this member function multiple
```

----
|Returns a list of all nodes in the graph which have no dependencies.

|
[source,c++]
----
size_t get_mem_use_current() const;
----
|Returns the size in bytes of the physical memory currently being used by this

> graph for <<graph-memory-allocations, graph-owned memory allocations>>.

Reviewer: Would say that "current" is implicit and we could shorted this to `get_mem_use".

Also thinking if "used" is too generic a description,  as it's scope could be interpreted as larger than just the memory allocated by the graph. `get_mem_alloc_size()` or something that references allocation explicitly might be better.

Author: Works for me, I've updated this to `get_mem_alloc_size()` now and adjusted the description to reflection that.

----
|Returns a list of all nodes in the graph which have no dependencies.

|
[source,c++]
----
size_t get_mem_alloc_size() const;
----

> |Returns the size in bytes of the physical memory allocated by this

Reviewer 2: I think this should have different behaviour for a modifiable graph and a finalised graph, as the finalised graph can also incorporate optimisations.

Reviewer 2: I think we should also have queries on the context for the amount of memory reserved and used for the whole device, something like `get_reserved_size_current` and `get_used_size_current`

Author: Currently this function is only available for graphs in the executable state, so it cannot be called on a modifiable graph. I think having it available for both is less relevant though. Using graph allocations creates a much more strict 1:1 relationship between an executable graph and a modifiable graph (because it can only be finalized once), and I think queries on the executable graph are more valuable because it can actually be used for execution.

----
|Returns a list of all nodes in the graph which have no dependencies.

|
[source,c++]
----

> size_t get_mem_alloc_size() const;

Reviewer 2: I think we should consider renaming this to something like `get_required_mem_size`, to avoid it being interpreted as the amount which is currently allocated by the memory pool.
